### PR TITLE
DeepONet Project Update: Model Accessibility and Checkpoint Restoration Fix

### PR DESCRIPTION
**Pull request by parsa-rahbari-82**
### Description:  
This pull request provides a series of important updates to the DeepONet project, primarily focusing on improving model accessibility and bug fixes. The following changes have been made:

1. Command-Line Model Parameters:  
   - Added command-line arguments to give users more control over the model configuration without needing to modify the source code directly.  
   - Users can now adjust critical parameters such as learning rate, number of sensors, and training epochs via the command line using Python's argparse library.  
   
   Example usage:  
   
   `python deeponet_pde.py -p ode -m 200 --num-train 10000 --num-test 3000--lr 1e-3 --epochs 50000 --nn deeponet --activation relu --init Glorot --stacked False`
   

2. Checkpoint Restoration Bug Fix:  
   - Resolved a bug related to restoring checkpoints, which was caused by an incorrect file path.  
   - The model can now properly restore from saved states for both training and inference.

3. Legendre Transform Flexibility:  
   - Replaced the hardcoded time value (T=2) with a dynamic time variable.  
   - This update ensures that the model can work with any time interval, providing flexibility and preventing extrapolation errors.

4. Compatibility with DeepONet’s Dynamic Systems:  
   - Updated the structure of X_train and ensured proper formatting for X_test to be compatible with TensorFlow's compat.v1 backend.  
   - Added necessary changes to avoid type mismatch errors, such as casting arrays in X_test to np.float32.

   Example code snippet:  
   
`   X_test = tuple([arr.astype(dtype=np.float32) for arr in X_test])
`   

### Testing:  
These changes have been thoroughly tested and confirmed to work reliably with TensorFlow's compat.v1 backend.

MIND Lab:
- Supervisor: Dr. Mostafa Abbaszadeh
Team:
   - AmirAli Fakhari Zavareh
   - Melika Taheri Soufi
   - Parsa Rahbari

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,181 @@
+### Python ###
```
Comment (lululxvi) [inline .gitignore]: Don't upload this file.

File: src/deeponet_pde.py [hunk 2]
```diff
@@ -1,16 +1,15 @@
-from __future__ import absolute_import
-from __future__ import division
-from __future__ import print_function
+from __future__ import absolute_import, division, print_function
```
Comment (lululxvi) [inline src/deeponet_pde.py]: This import line is not needed anymore for python 3.

File: src/system.py [hunk 3]
```diff
@@ -1,36 +1,34 @@
-from __future__ import absolute_import
-from __future__ import division
-from __future__ import print_function
+from __future__ import absolute_import, division, print_function
 
+import config
```
Comment (lululxvi) [inline src/system.py]: Use the original order of import, i.e., import python libraries first, and then self-defined files.

**Comment by parsa-rahbari-82** [comment]
Dear Dr. Lu 
Thanks for the review! I’ve addressed the points you raised:

- Redundant python 2 removed.
- .gitignore file removed.
- Using original order of import.

File: src/spaces.py [hunk 4]
```diff
@@ -1,15 +1,11 @@
...
 import matplotlib.pyplot as plt
 import numpy as np
 from pathos.pools import ProcessPool
-from scipy import linalg, interpolate
+from scipy import interpolate, linalg
 from sklearn import gaussian_process as gp
 
-import config
 from utils import eig
+import config
```
Comment (lululxvi) [inline src/spaces.py]: `import ...` first, and then `from ... import ...`

**Comment by parsa-rahbari-82** [comment]
I reordered imports according to PEP 8 guidelines.

File: requirements.txt
Comment (lululxvi) [inline requirements.txt:1]: Why revise this file?

File: .gitignore
Comment (lululxvi) [inline .gitignore:1]: No need to add this file. Different users may have different style.

File: src/test/u_xsin2x.dat
Comment (lululxvi) [inline src/test/u_xsin2x.dat:1]: What is this file?

File: requirements.txt [hunk 5]
```diff
@@ -1,8 +0,0 @@
-deepxde
```
Comment (lululxvi) [inline requirements.txt]: Keep this file as the original one.

Comment (lululxvi) [reply requirements.txt]: @AmirAliFakhari It seems the file roolls back to the wrong version.

Comment (AmirAliFakhari) [reply requirements.txt]: Dear Mr. Lu Lu,

I hope you're doing well. I've been working on the DeepONet project for the past few months and truly appreciate your feedback throughout the process.

To help streamline the next steps and avoid back-and-forth revisions, would it be possible for you to share a complete list of the issues or changes you'd like to see in the current version? This would really help me address everything thoroughly and ensure that the final result meets your expectations.

File: src/deeponet_pde.py [hunk 6]
```diff
@@ -1,16 +1,13 @@
-from __future__ import absolute_import
-from __future__ import division
-from __future__ import print_function
-
+import argparse
 import itertools
-
+import deepxde as dde
 import numpy as np
 import tensorflow as tf
```
Comment (lululxvi) [inline src/deeponet_pde.py:5]:
```
import argparse
import deepxde as dde
import itertools
import numpy as np
import tensorflow as tf
```

File: src/deeponet_pde.py [hunk 7]
```diff
@@ -1,16 +1,13 @@
...
 
-import deepxde as dde
-from spaces import FinitePowerSeries, FiniteChebyshev, GRF
-from system import LTSystem, ODESystem, DRSystem, CVCSystem, ADVDSystem
-from utils import merge_values, trim_to_65535, mean_squared_error_outlier, safe_test
+from spaces import GRF, FiniteChebyshev, FinitePowerSeries
```
Comment (lululxvi) [inline src/deeponet_pde.py:7]:
```
from spaces import FiniteChebyshev, FinitePowerSeries, GRF
```

File: src/deeponet_pde.py [hunk 8]
```diff
@@ -21,9 +18,10 @@ def test_u_lt(nn, system, T, m, model, data, u, fname):
     ns = np.arange(system.npoints_output)[:, None]
     X_test = [np.tile(sensor_value, (system.npoints_output, 1)), ns]
     y_test = s
-    if nn != "opnn":
+    if nn != "deeponet":
         X_test = merge_values(X_test)
-    y_pred = model.predict(data.transform_inputs(X_test))
+    X_test = tuple([arr.astype(dtype=np.float32) for arr in X_test])
+    y_pred = model.predict(X_test)
```
Comment (lululxvi) [inline src/deeponet_pde.py:24]: Why no `data.transform_inputs`? This is for data normalization.

File: src/deeponet_pde.py [hunk 9]
```diff
@@ -103,11 +109,11 @@ def g(s, u, x):
         # Nonlinear ODE
         # return -s**2 + u
         # Gravity pendulum
-        # k = 1
-        # return [s[1], - k * np.sin(s[0]) + u]
+        k = 1
```
Comment (lululxvi) [inline src/deeponet_pde.py:112]: There is `return` in line 108. The code here is not used.

File: src/spaces.py [hunk 10]
```diff
@@ -1,14 +1,10 @@
...
 import numpy as np
+import config
+
 from pathos.pools import ProcessPool
-from scipy import linalg, interpolate
+from scipy import interpolate, linalg
 from sklearn import gaussian_process as gp
-
-import config
 from utils import eig
```
Comment (lululxvi) [inline src/spaces.py:8]:
```
import config

import matplotlib.pyplot as plt
...
```

File: src/system.py [hunk 11]
```diff
@@ -1,36 +1,34 @@
-from __future__ import absolute_import
-from __future__ import division
-from __future__ import print_function
-
 import numpy as np
+import config
```
Comment (lululxvi) [inline src/system.py:2]: `config` is a local file. Move it below as the original code.

File: src/utils.py [hunk 12]
```diff
@@ -1,12 +1,9 @@
-from __future__ import absolute_import
-from __future__ import division
-from __future__ import print_function
-
 import sys
 import time
+import numpy as np
```
Comment (lululxvi) [inline src/utils.py:3]: Use the original import format for `numpy`
